drivers/misc/cldatadrv/rcar_cluster_drv.c

```diff
@@ -0,0 +1,330 @@
+					complete(&event->completed);
+				} else {
+					event->ack_received = 1;
+					complete(&event->ack);
+				}
+			}
+		}
+		read_unlock(&clusterdrv->event_list_lock);
+	}
+	return 0;
+}
+
+static int rpmsg_cluster_probe(struct rpmsg_device* rpdev)
+{
+	rcar_cluster_device_t *clusterdvc = NULL;
+	int ret = 0;
+	struct miscdevice *dev = NULL;
+	taurus_cluster_res_msg_t res_msg;
+
+	taurus_cluster_data_t cluster_data = {
+		.value = 10,
+		.ioctl_cmd = 1,
+	};
+	dev_info(&rpdev->dev, "cluster: send_msg: probe\n");
+
+	clusterdvc = kzalloc(sizeof(*clusterdvc), GFP_KERNEL);
+
+	if (clusterdvc== NULL)
+		return -ENOMEM;
+	
+	clusterdvc->rpdev = rpdev;
+	dev = &clusterdvc->dev;
+	dev->parent = &rpdev->dev;
+	dev->minor = MISC_DYNAMIC_MINOR;
+
+	mutex_lock(&idr_lock);
+	ret = idr_alloc(&misc_idr, clusterdvc, 0, MAX_MISC_RPMSG, GFP_KERNEL);
+	mutex_unlock(&idr_lock);
+
+	if (ret < 0) {
+		goto free_cluster_data;
+	}
+
+	dev->name = kstrdup("cluster-taurus", GFP_KERNEL);
+	if (!dev->name) {
+		ret = -ENOMEM;
+		goto free_name;
+	}
+
+	dev->fops = &misc_fops;
+	ret = misc_register(dev);
+	
+	dev_set_drvdata(&rpdev->dev, clusterdvc);
+	
+	INIT_LIST_HEAD(&clusterdvc->taurus_event_list_head);
+	rwlock_init(&clusterdvc->event_list_lock);
+	/*send a ping of message with dummy data*/
+	send_msg(rpdev, &cluster_data, &res_msg);
+
+	return ret;
+
+free_name:
+	kfree(dev->name);
+	dev->name = NULL;
+
+free_cluster_data:
+	kfree(clusterdvc);
+
+	return ret;
+}
+
+static void rpmsg_cluster_remove(struct rpmsg_device *rpdev)
+{
+	rcar_cluster_device_t *data = dev_get_drvdata(&rpdev->dev);
+	
+	mutex_lock(&idr_lock);
+	idr_remove(&misc_idr, 0);
+	mutex_unlock(&idr_lock);
+	
+	misc_deregister(&data->dev);
+	
+	kfree(data->dev.name);
+	data->dev.name = NULL;
+
+	kfree(data);
+
+    return;
+}
+
+static int __init cluster_drv_init(void)
+{
+	int ret = 0;
+
+	ret = register_rpmsg_driver(&rpmsg_cluster_drv);
+	if (ret < 0) {
+		pr_err("failed to register cluster_drv_init driver\n");
+		return -EAGAIN;
+	}
+
+	return ret;
+}
+/*postcore_initcall*/
+late_initcall(cluster_drv_init);
+
+static void __exit cluster_drv_exit(void)
+{
+	unregister_rpmsg_driver(&rpmsg_cluster_drv);
+}
+
+static int misc_open(struct inode *inode, struct file *p_file)
+{
+	rcar_cluster_device_t * cldev = idr_find(&misc_idr, 0);
+	p_file->private_data = cldev;
+
+	return 0;
+}
+
+static int misc_release(struct inode *inode, struct file *p_file)
+{
+	return 0;
+}
+
+static ssize_t misc_write(struct file *p_file,
+		     const char __user *buf,
+		     size_t count,
+		     loff_t *f_pos)
+{
+	rcar_cluster_device_t * cldev = p_file->private_data;
+	void *kbuf;
+	int ret;
+	taurus_cluster_data_t * data = NULL;
+	taurus_cluster_res_msg_t res;
+
+	kbuf = kzalloc(count, GFP_KERNEL);
+	if (!kbuf)
+		return -ENOMEM;
+
+	if (copy_from_user(kbuf, buf, count)) {
+		ret = -EFAULT;
+		goto free_kbuf;
+	}
+	
+	data = (taurus_cluster_data_t*)kbuf;
```

Comment:
If userspace try to wrote exactly 1 byte (or 2, or 3 bytes), you will have big problems.
This is one of the reasons why you need to use IOCTLs.